(DOCSP-17707): Update code examples and text to better illustrate offline vs online opening a synced realm (#1247)
mongodb/docs-realm: source/sdk/ios/examples/sync-changes-between-devices.txt

~~~~~~~~~~~~~~~~~~~

The first time a user opens a {+realm+}, you'll need to log the user in
and download a {+realm+} before opening it.

Upon this initial login, we cache login credentials, so you can open new 
{+realms+} after checking for a logged-in user. 

Log In and Download a Synced Realm
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabs-realm-languages::
   
   .. tab::
      :tabid: swift

      The first time you log in and open a synced {+realm+}, you'll log in the
      user, and pass the  user's :swift-sdk:`Configuration <Structs/Realm/Configuration.html>` 
      object with the desired :ref:`partition <partitioning>` value to
      :swift-sdk:`Realm.asyncOpen(configuration:)
      <Structs/Realm.html#/s:10RealmSwift0A0V9asyncOpen13configuration13callbackQueue0F0AC05AsyncD4TaskVAC13ConfigurationV_So17OS_dispatch_queueCyACSg_s5Error_pSgtctFZ>`:

      When this is the first time opening the {+realm+} on the device, ``asyncOpen`` 
      downloads the entire remote {+realm+} before opening the file on device. 
      If a {+realm+} is already present on the device, ``asyncOpen`` downloads 
      only the latest changes since the {+realm+} last synced with the {+app+}.
      

>       A common pattern is to use ``asyncOpen`` during a login flow, but use 

Reviewer: ```suggestion
      A common pattern uses ``asyncOpen`` during a login flow, but use 
```
passive -> active voice reword, though I think in this case passive voice isn't a huge problem.

      :objc-sdk:`RLMRealmConfiguration
      <Classes/RLMRealmConfiguration.html>` object with the desired
      :ref:`partition <partitioning>` value to :objc-sdk:`+[RLMRealm
      asyncOpenWithConfiguration:callbackQueue:callback]
      <Classes/RLMRealm.html#/c:objc(cs)RLMRealm(cm)asyncOpenWithConfiguration:callbackQueue:callback:>`:

      If this is the first time opening the {+realm+} on the device, ``asyncOpen`` 
      downloads the entire remote {+realm+} before opening the file on device. 
      If a {+realm+} is already present on the device, ``asyncOpen`` downloads 
      only the latest changes since the {+realm+} last synced with the {+app+}.
      
      A common pattern is to use ``asyncOpen`` during a login 
      flow, but use ``init`` to open a {+realm+} on subsequent opens.

      .. literalinclude:: /examples/generated/code/start/Sync.codeblock.open-synced-realm.m
         :language: objectivec

Open a Synced Realm that has Previously Been Synced
---------------------------------------------------

After you have authenticated a user and downloaded a synced {+realm+},
you can open the same {+realm+} again without re-authenticating. We cache

> credentials until you :swift-sdk:`logout the user <Extensions/User.html#/s:So7RLMUserC10RealmSwiftE6logOut7Combine6FutureCyyts5Error_pGyF>`, 

Reviewer: *I'm fairly certain cached credentials expire at some point but I haven't gotten concrete details on when or how

or :ref:`perform a client reset <ios-client-reset>`. 

With a synced {+realm+} on the device and a previously logged-in user, you
can choose to open the synced {+realm+} in two ways:

- :ref:`Open a synced {+realm+} with local data <ios-open-synced-realm-with-local-data>` 
  immediately. This works regardless of network status, and enables the 
  user to start working with the {+realm+} without waiting for changes to 
  download from the server. However, data that is on the device may be stale. 
  Additionally, this may result in behavior that seems unexpected or surprising 

>   to the user. Data that has synced in the background may appear "suddenly" 

Reviewer: ```suggestion
  to the user. Data may appear "suddenly" as it syncs in the background
```
Passive -> active reword

Open a Synced Realm
-------------------

The typical flow for opening a synced {+realm+} for the first time involves:

1. Authenticating the user
#. Creating a :swift-sdk:`sync configuration <Extensions/User.html#/s:So7RLMUserC10RealmSwiftE13configuration14partitionValueAC0B0V13ConfigurationVx_tAC4BSONRzlF>`
#. Opening the user's synced {+realm+} with the configuration.

At authentication, we cache user credentials in a ``sync_metadata.realm`` 
file on device.

When you open a synced {+realm+} after authenticating, you can bypass the 
login flow and go directly to opening the synced {+realm+}, using the same 
sync configuration you already created.

With cached credentials, you can:

- Open a synced {+realm+} immediately with the data that is on the device.
  You can use this method offline or online.
- Open a synced {+realm+} after syncing changes with your {+app+}. You can
  only use this method if the user is online.

.. important::

>    {+sync+} does not currently support watchOS.

Reviewer: I can't help but think this note should live somewhere else now, maybe on the iOS landing page section that discusses sync?

- Some methods of opening synced {+realms+} require users to be online, 
  while local {+realms+} can be used entirely offline.

You can copy data from a :ref:`local {+client-database+} <ios-open-a-local-realm>` 
to a synced {+realm+}, but you cannot sync a local {+client-database+}. You
must initialize a synced {+realm+} with a :swift-sdk:`sync configuration <Extensions/User.html#/s:So7RLMUserC10RealmSwiftE13configuration14partitionValueAC0B0V13ConfigurationVx_tAC4BSONRzlF>`. 

Open a Synced Realm for the First Time
--------------------------------------

Authenticate a User
~~~~~~~~~~~~~~~~~~~

The first time a user opens a {+realm+}, you'll need to authenticate the user. 
Upon this initial login, we cache login credentials. On subsequent opens, you 
can check for a logged-in user, and then open new {+realms+}. You don't need to
complete the login flow while you have a logged-in user. 

Depending on your business logic, you may also require the user to download 
data from {+backend+} before opening a synced {+realm+} on device.

Log In and Initialize a Synced Realm
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabs-realm-languages::
   
   .. tab::

>       :tabid: swift

Reviewer: What's going on with the tabbing here? Is there no objective-c equivalent?

Sidenote: a subsection title + tab combo generates an enormous gap between those elements and it looks totally awful. Maybe worth creating a platform ticket?

      The first time you log in and open a synced {+realm+}, you'll log in the
      user, and pass the  user's :swift-sdk:`Configuration <Structs/Realm/Configuration.html>` 
      object with the desired :ref:`partition <partitioning>` value to
      :swift-sdk:`Realm.asyncOpen(configuration:)
      <Structs/Realm.html#/s:10RealmSwift0A0V9asyncOpen13configuration13callbackQueue0F0AC05AsyncD4TaskVAC13ConfigurationV_So17OS_dispatch_queueCyACSg_s5Error_pSgtctFZ>`:

      This opens a synced {+realm+} on the device with data that is present
      on the device. The {+realm+} will attempt to sync with your {+app+} in
      the background to check for changes on the server, or upload changes 
      that the user has made.

      .. literalinclude:: /examples/generated/code/start/Sync.codeblock.login-and-init-synced-realm.swift
         :language: swift

> Log In and Download a Synced Realm

Reviewer: There's something about this title (and the preceding subsection title) that really confuses me. I think we should really consider renaming these subsections, because "initialize" and "download" are being used to convey *extremely specific* realm-related concepts here that I don't think land fully with the average reader (or even myself!)

Reviewer: Maybe we can just use the same terminology as the next section uses, "with Data on Device" and "After Downloading Changes"?

  This requires the user to have a network connection, so you should 
  :ref:`check the network connection <ios-check-network-connection>` before 
  attempting to open the {+realm+} from the server. This method of opening the 
  {+realm+} ensures that the user always has the most up-to-date data from the 
  server. The cost of using this method is an upfront pause while loading, and 
  being unable to open the {+realm+} offline.

.. _ios-open-synced-realm-with-data-on-device:

Open a Synced Realm with Data on Device
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can use initializers to open the {+realm+} immediately, using the data
that is on the device. If you want to open a synced {+realm+} offline, use 
this method. You must have a cached, logged-in user.

.. example::

   Consider the example of a note-taking app. You might decide it's most 
   important for your user to be able to quickly jot down a note. The user 
   shouldn't have to wait while downloading changes that a family member made 
   to a shared note. In this case, opening a {+realm+} with ``init`` gets the 
   user to the UI right away, or while the user is offline. When the user has

>    a network connection, changes will sync in the background.

Reviewer: I wonder if a counterexample of the *opposite case* could be helpful here too, to contrast the differences

- :ref:`Open a synced {+realm+} with data that is on the device 
  <ios-open-synced-realm-with-data-on-device>` immediately. This works regardless 
  of network status, and enables the user to start working with the {+realm+} 
  without waiting for changes to download from the server. However, data that 
  is on the device may be stale. Additionally, this may result in behavior that 
  seems unexpected or surprising to the user. Data may appear "suddenly" as 
  it syncs in the background after the user has already begun working with the 
  {+realm+}. You'll need to account for this in your UI logic.
- :ref:`Open a synced {+realm+} after syncing changes <ios-open-synced-realm-after-sync>`. 
  This requires the user to have a network connection, so you should 
  :ref:`check the network connection <ios-check-network-connection>` before 
  attempting to open the {+realm+} from the server. This method of opening the 
  {+realm+} ensures that the user always has the most up-to-date data from the 
  server. The cost of using this method is an upfront pause while loading, and 
  being unable to open the {+realm+} offline.

.. _ios-open-synced-realm-with-data-on-device:

> Open a Synced Realm with Data on Device

Reviewer: Very very very small nitpick: the other header uses the phrasing "the Device", this uses "Device" without an article. Can we standardize that?

      .. literalinclude:: /examples/generated/code/start/Sync.codeblock.open-synced-realm-synchronously.m
         :language: objectivec

.. _ios-open-synced-realm-from-server:

Open a Synced Realm from the Server
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tabs-realm-languages::
   
   .. tab::
      :tabid: swift

>       To open a synced {+realm+} from the server, pass the logged-in user's

Reviewer: I think this wording is a little misleading (or at least it feels that way to me) -- "from the server" makes it sound as if we're talking about opening the realm on the backend, whereas I think you're trying to convey "with initial data loaded from the server" or something along those lines. Consider rewording this sentence and the subsection heading to make that distinction a little clearer.
